### PR TITLE
Expose Vertx core metrics in Prometheus format

### PR DESCRIPTION
**Pull request by pierDipi**
This PR exposes [Vertx core metrics](https://vertx.io/docs/vertx-micrometer-metrics/java/#_vert_x_core_tools_metrics) in Prometheus format and add
notes on how to configure Prometheus to scrape data plane pods.

Spec metrics in a separate PR.

Signed-off-by: Pierangelo Di Pilato <pierangelodipilato@gmail.com>

Part of #227 

## Proposed Changes

- Expose Vertx core metrics in Prometheus format

## Prometheus dashboard

image

**Release Note**

```release-note
- 🎁 Add new feature
Data plane pods expose server and client metrics in Prometheus format.
```

### CODE REVIEW

File: INTERNALS.md [hunk 1]
```diff
@@ -38,3 +38,64 @@ following the
...
+      - source_labels: [__meta_kubernetes_namespace, __meta_kubernetes_pod_label_app, __meta_kubernetes_pod_container_port_name]
+        action: keep
+        regex: knative-eventing;kafka-sink-receiver;http-metrics
+      # Rename metadata labels to be reader friendly
+      - source_labels: [__meta_kubernetes_namespace]
+        target_label: namespace
+      - source_labels: [__meta_kubernetes_pod_name]
+        target_label: pod
+      - source_labels: [__meta_kubernetes_service_name]
+        target_label: service
```
Comment (mattmoor) [inline INTERNALS.md:100]: Format markdown:
```suggestion
# scrape kafka-broker-receiver
- job_name: kafka-broker-receiver
  scrape_interval: 5m
  scrape_timeout: 30s
  kubernetes_sd_configs:
    - role: pod
  relabel_configs:
    # Scrape only the the targets matching the following metadata
    - source_labels:
        [
          __meta_kubernetes_namespace,
          __meta_kubernetes_pod_label_app,
          __meta_kubernetes_pod_container_port_name,
        ]
      action: keep
      regex: knative-eventing;kafka-broker-receiver;http-metrics
    # Rename metadata labels to be reader friendly
    - source_labels: [__meta_kubernetes_namespace]
      target_label: namespace
    - source_labels: [__meta_kubernetes_pod_name]
      target_label: pod
    - source_labels: [__meta_kubernetes_service_name]
      target_label: service
# scrape kafka-broker-dispatcher
- job_name: kafka-broker-dispatcher
  scrape_interval: 5m
  scrape_timeout: 30s
  kubernetes_sd_configs:
    - role: pod
  relabel_configs:
    # Scrape only the the targets matching the following metadata
    - source_labels:
        [
          __meta_kubernetes_namespace,
          __meta_kubernetes_pod_label_app,
          __meta_kubernetes_pod_container_port_name,
        ]
      action: keep
      regex: knative-eventing;kafka-broker-dispatcher;http-metrics
    # Rename metadata labels to be reader friendly
    - source_labels: [__meta_kubernetes_namespace]
      target_label: namespace
    - source_labels: [__meta_kubernetes_pod_name]
      target_label: pod
    - source_labels: [__meta_kubernetes_service_name]
      target_label: service
# scrape kafka-sink-receiver
- job_name: kafka-sink-receiver
  scrape_interval: 5m
  scrape_timeout: 30s
  kubernetes_sd_configs:
    - role: pod
  relabel_configs:
    # Scrape only the the targets matching the following metadata
    - source_labels:
        [
          __meta_kubernetes_namespace,
          __meta_kubernetes_pod_label_app,
          __meta_kubernetes_pod_container_port_name,
        ]
      action: keep
      regex: knative-eventing;kafka-sink-receiver;http-metrics
    # Rename metadata labels to be reader friendly
    - source_labels: [__meta_kubernetes_namespace]
      target_label: namespace
    - source_labels: [__meta_kubernetes_pod_name]
      target_label: pod
    - source_labels: [__meta_kubernetes_service_name]
      target_label: service
```

**Review by mattmoor — COMMENTED** [review]
Produced via:
  `prettier --write --prose-wrap=always $(find -name '*.md' | grep -v vendor | grep -v .github | grep -v docs/cmd/)`

**Comment by pierDipi** [comment]
Waiting for the release!
/hold

File: INTERNALS.md (hunk 1 above)
Comment (mattmoor) [inline INTERNALS.md:100]: Format markdown:
```suggestion
# scrape kafka-broker-receiver
- job_name: kafka-broker-receiver
  scrape_interval: 5m
  scrape_timeout: 30s
  kubernetes_sd_configs:
    - role: pod
  relabel_configs:
    # Scrape only the the targets matching the following metadata
    - source_labels:
        [
          __meta_kubernetes_namespace,
          __meta_kubernetes_pod_label_app,
          __meta_kubernetes_pod_container_port_name,
        ]
      action: keep
      regex: knative-eventing;kafka-broker-receiver;http-metrics
    # Rename metadata labels to be reader friendly
    - source_labels: [__meta_kubernetes_namespace]
      target_label: namespace
    - source_labels: [__meta_kubernetes_pod_name]
      target_label: pod
    - source_labels: [__meta_kubernetes_service_name]
      target_label: service
# scrape kafka-broker-dispatcher
- job_name: kafka-broker-dispatcher
  scrape_interval: 5m
  scrape_timeout: 30s
  kubernetes_sd_configs:
    - role: pod
  relabel_configs:
    # Scrape only the the targets matching the following metadata
    - source_labels:
        [
          __meta_kubernetes_namespace,
          __meta_kubernetes_pod_label_app,
          __meta_kubernetes_pod_container_port_name,
        ]
      action: keep
      regex: knative-eventing;kafka-broker-dispatcher;http-metrics
    # Rename metadata labels to be reader friendly
    - source_labels: [__meta_kubernetes_namespace]
      target_label: namespace
    - source_labels: [__meta_kubernetes_pod_name]
      target_label: pod
    - source_labels: [__meta_kubernetes_service_name]
      target_label: service
# scrape kafka-sink-receiver
- job_name: kafka-sink-receiver
  scrape_interval: 5m
  scrape_timeout: 30s
  kubernetes_sd_configs:
    - role: pod
  relabel_configs:
    # Scrape only the the targets matching the following metadata
    - source_labels:
        [
          __meta_kubernetes_namespace,
          __meta_kubernetes_pod_label_app,
          __meta_kubernetes_pod_container_port_name,
        ]
      action: keep
      regex: knative-eventing;kafka-sink-receiver;http-metrics
    # Rename metadata labels to be reader friendly
    - source_labels: [__meta_kubernetes_namespace]
      target_label: namespace
    - source_labels: [__meta_kubernetes_pod_name]
      target_label: pod
    - source_labels: [__meta_kubernetes_service_name]
      target_label: service
```

**Review by mattmoor — COMMENTED** [review]
Produced via:
  `prettier --write --prose-wrap=always $(find -name '*.md' | grep -v vendor | grep -v .github | grep -v docs/cmd/)`

File: data-plane/dispatcher/pom.xml [hunk 2]
```diff
@@ -29,6 +29,23 @@
...
+          <groupId>org.hdrhistogram</groupId>
+          <artifactId>HdrHistogram</artifactId>
+        </exclusion>
+      </exclusions>
+    </dependency>
+
+    <dependency>
+      <groupId>io.micrometer</groupId>
+      <artifactId>micrometer-core</artifactId>
+      <version>${micrometer.version}</version>
```
Comment (slinkydeveloper) [inline data-plane/dispatcher/pom.xml:43]: You don't need the version

File: data-plane/dispatcher/pom.xml [hunk 3]
```diff
@@ -29,6 +29,23 @@
...
+    </dependency>
+    <dependency>
+      <groupId>io.micrometer</groupId>
+      <artifactId>micrometer-registry-prometheus</artifactId>
+      <version>${micrometer.version}</version>
```
Comment (slinkydeveloper) [inline data-plane/dispatcher/pom.xml:48]: Same

File: data-plane/pom.xml [hunk 4]
```diff
@@ -73,14 +74,45 @@
...
+        <type>pom</type>
+        <scope>import</scope>
+      </dependency>
+
+      <!-- Micrometer -->
+      <dependency>
+        <groupId>io.micrometer</groupId>
+        <artifactId>micrometer-core</artifactId>
+        <version>${micrometer.version}</version>
+        <type>pom</type>
```
Comment (slinkydeveloper) [inline data-plane/pom.xml:104]: You don't need the type pom scope import, that' only for boms

Comment (slinkydeveloper) [reply data-plane/pom.xml:104]: More than manually importing the core and the registry prometheus, why don't you import the bom? https://mvnrepository.com/artifact/io.micrometer/micrometer-bom/1.5.5

File: data-plane/core/src/main/java/dev/knative/eventing/kafka/broker/core/utils/BaseEnv.java [hunk 5]
```diff
@@ -4,15 +4,26 @@
...
+
   private final String producerConfigFilePath;
   private final String dataPlaneConfigFilePath;
 
+  private final int metricsPort;
+  private final String metricsPath;
+  private final boolean metricsPublishQuantiles;
+
   public BaseEnv(Function<String, String> envProvider) {
+    this.metricsPath = requireNonNull(envProvider.apply(METRICS_PATH));
```
Comment (slinkydeveloper) [inline data-plane/core/src/main/java/dev/knative/eventing/kafka/broker/core/utils/BaseEnv.java:24]: Can we have the metrics configuration optional? If you don't add the envs for that, the metrics options are just not created and the metrics listener doesn't start. WDYT?

File: data-plane/pom.xml [hunk 6]
```diff
@@ -73,14 +74,30 @@
...
     <dependencies>
+      <!-- Vertx -->
       <dependency>
         <groupId>io.vertx</groupId>
         <artifactId>vertx-stack-depchain</artifactId>
         <version>${vertx.version}</version>
         <type>pom</type>
         <scope>import</scope>
       </dependency>
+      <dependency>
```
Comment (slinkydeveloper) [inline data-plane/pom.xml:85]: This should be included in the vertx bom

File: data-plane/core/src/main/java/dev/knative/eventing/kafka/broker/core/utils/BaseEnv.java (hunk 5 above)
Comment (pierDipi) [reply data-plane/core/src/main/java/dev/knative/eventing/kafka/broker/core/utils/BaseEnv.java:24]: It makes sense, however, I would like to have a more concise way of disabling metrics through a ConfigMap regardless of the env variable's values. 
The main reason for this is that to disable metrics users need to change every deployment, `kafka-broker-receiver`, `kafka-broker-dispatcher`, and `kafka-sink-receiver` (and more are coming), and that doesn't seem optimal to me. 

The `ConfigMap` format can be:
```yaml
metrics.enabled: (true|false)
```
or 
```properties
metrics.enabled=(true|false)
```

WDYT? Should we add a `ConfigMap`?

Comment (slinkydeveloper) [reply data-plane/core/src/main/java/dev/knative/eventing/kafka/broker/core/utils/BaseEnv.java:24]: Do we have some "monitoring" config map in eventing that we can use to check if metrics are enabled or not?

Comment (pierDipi) [reply data-plane/core/src/main/java/dev/knative/eventing/kafka/broker/core/utils/BaseEnv.java:24]: No, `config-observability` hasn't any `enable`/`disable` flag for metrics:

```yaml
    # metrics.backend-destination field specifies the system metrics destination.
    # It supports either prometheus (the default) or stackdriver.
    # Note: Using stackdriver will incur additional charges
    metrics.backend-destination: prometheus

    # metrics.request-metrics-backend-destination specifies the request metrics
    # destination. If non-empty, it enables queue proxy to send request metrics.
    # Currently supported values: prometheus, stackdriver.
    metrics.request-metrics-backend-destination: prometheus

    # metrics.stackdriver-project-id field specifies the stackdriver project ID. This
    # field is optional. When running on GCE, application default credentials will be
    # used if this field is not provided.
    metrics.stackdriver-project-id: "<your stackdriver project id>"

    # metrics.allow-stackdriver-custom-metrics indicates whether it is allowed to send metrics to
    # Stackdriver using "global" resource type and custom metric type if the
    # metrics are not supported by "knative_broker", "knative_trigger", and "knative_source" resource types.
    # Setting this flag to "true" could cause extra Stackdriver charge.
    # If metrics.backend-destination is not Stackdriver, this is ignored.
    metrics.allow-stackdriver-custom-metrics: "false"

    # profiling.enable indicates whether it is allowed to retrieve runtime profiling data from
    # the pods via an HTTP server in the format expected by the pprof visualization tool. When
    # enabled, the Knative Eventing pods expose the profiling data on an alternate HTTP port 8008.
    # The HTTP context root for profiling is then /debug/pprof/.
    profiling.enable: "false"

    # sink-event-error-reporting.enable whether the adapter reports a kube event to the CRD indicating
    # a failure to send a cloud event to the sink.
    sink-event-error-reporting.enable: "false"
```

We can propose to add it, it requires a bit of work to implement it tho.

Comment (slinkydeveloper) [reply data-plane/core/src/main/java/dev/knative/eventing/kafka/broker/core/utils/BaseEnv.java:24]: well, let me ask another question then, is it possible in eventing to disable metrics collection? If not, then we don't allow disabling them too and that's it

Comment (pierDipi) [reply data-plane/core/src/main/java/dev/knative/eventing/kafka/broker/core/utils/BaseEnv.java:24]: I think it's not possible ATM.

**Pull request merged**